src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -55,6 +55,28 @@ public static Task WriteMessageAsync(this PipeWriter pipeWriter, byte[] messageD
                 return Task.FromException(new RpcException(SendingMessageExceedsLimitStatus));
             }
 
+            // Must call StartAsync before the first pipeWriter.GetSpan() in WriteHeader
+            var response = serverCallContext.HttpContext.Response;
+            if (!response.HasStarted)
+            {
```

Comment:
It might be cleaner to remove the above condition since that same condition is at the start of Response.StartAsync(). When the response has already started, StartAsync will immediately return a static completed task.